### PR TITLE
Fix Payment block error while editing the blocks checkout page.

### LINKED ISSUE #8676: [GlobalStep] "This block has encountered an error and cannot be previewed." message displayed "Checkout" block editor page after enabling "WooPayments" option.
### Bug Description:
"This block has encountered an error and cannot be previewed." message displayed "Checkout" block editor page after enabling "WooPayments" option from "Woocommerce > Settings > Payments" page. 

### Environment:
Woocommerce Version: 8.9.0.20
WooPayments version: 7.5.1
WordPress version: 6.5.2

PC:
Windows 10
Chrome (Version 124.0.6367.61)
Firefox (Version 125.0.1)

### Steps To Reproduce:
1. Create a test site.
2. Install and activate all the required plugins.
3. Install and activate the Woocommerce version 8.9.0.20 plugins.
4. Install and set up WooCommerce Payments in test mode.
5. Go to "Woocommerce > Settings > Payments > Payment Methods" page.
6. Disable "WooPayments" option.
7. Click "Pages > Add New Page" and create a new test page.
8. Add "Checkout" block using inserter.
9. Scroll down to Payment section and observe no error mesag eis displayed.
10. Now go to "Woocommerce > Settings > Payments > Payment Methods" page.
11. Enable "WooPayments" option.
12. Navigate to the checkout block page again created in step 7.
13. Observe that, "This block has encountered an error and cannot be previewed." message displayed "Checkout" block editor page after enabling "WooPayments" option 

### Actual Result:
"This block has encountered an error and cannot be previewed." message displayed "Checkout" block editor page after enabling "WooPayments" option from "Woocommerce > Settings > Payments" page. 

### Expected Result:
"Checkout" block editor page should be displayed properly without any error after enabling "WooPayments" option from "Woocommerce > Settings > Payments" page. 

### Video:

https://github.com/Automattic/woocommerce-payments/assets/41110392/138ccdd0-e08a-4883-aa87-a21024220cad

*Isolating the problem (mark completed items with an [x]):*
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

<details>

`
### WordPress Environment ###

WC Version: 8.9.0.20
REST API Version: ✔ 8.9.0.20
Action Scheduler Version: ✔ 3.7.4
Log Directory Writable: ✔
WP Version: 6.5.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 8.1.28
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 8.7.1
OpenSSL/3.0.11

SUHOSIN Installed: –
MySQL Version: 10.6.17-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.9.0.20
WC Database Prefix: wp_
Total Database Size: 3.81MB
Database Data Size: 1.91MB
Database Index Size: 1.90MB
wp_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.05MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_jetpack_sync_queue: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.30MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpml_mails: Data: 0.17MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 8
post: 2
product: 18
product_variation: 7
revision: 1
shop_coupon: 1
shop_order_placehold: 15
wp_font_face: 33
wp_font_family: 12
wp_global_styles: 1
wp_navigation: 1
wp_template: 1
wp_template_part: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (5) ###

Query Monitor: by John Blackbourn – 3.15.0
Code Snippets: by Code Snippets Pro – 3.6.4
WooPayments: by Automattic – 7.5.1
WooCommerce: by Automattic – 8.9.0.20
WP Mail Logging: by WP Mail Logging Team – 1.12.0

### Inactive Plugins (2) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.2
Jetpack: by Automattic – 13.4-a.3

### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php
db.php: Query Monitor Database Class (Drop-in)
object-cache.php: Memcached

### Must Use Plugins (1) ###

load.php: by  –

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 551 KB

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: #2 - /sample-page/

### Theme ###

Name: Twenty Twenty-Four
Version: 1.1
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/error.php
/srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/success.php

### WooPayments ###

Version: 7.5.1
Connected to WPCOM: Yes
WPCOM Blog ID: 232062259
Account ID: acct_1P6vCXC4BWlC9t2H
Payment Gateway: Enabled
Test Mode: Enabled
Enabled APMs: card
WooPay: Disabled
WooPay Incompatible Extensions: No
Apple Pay / Google Pay: Enabled (product,cart,checkout)
Fraud Protection Level: basic
Multi-currency: Enabled
Public Key Encryption: Disabled
Auth and Capture: Enabled
Documents: Disabled
Logging: Enabled

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
product-pre-publish-modal
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: minified-js
new-product-management-experience
product-custom-fields
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-04-19 13:33:32 +00:00
Options: ✔
Notes: 68
Onboarding: completed

### Action Scheduler ###

Canceled: 4
Oldest: 2024-04-18 14:26:50 +0000
Newest: 2024-04-18 15:05:40 +0000

Complete: 99
Oldest: 2024-04-18 13:33:34 +0000
Newest: 2024-04-19 11:54:14 +0000

Failed: 3
Oldest: 2024-04-18 13:35:54 +0000
Newest: 2024-04-18 13:44:35 +0000

Pending: 2
Oldest: 2024-04-19 13:34:39 +0000
Newest: 2024-04-19 23:50:01 +0000

### Status report information ###

Generated at: 2024-04-19 12:08:12 +00:00
`

</details>

### PR DESCRIPTION
**Pull request by brettshumaker**
Fixes #8676 

#### Changes proposed in this Pull Request
This PR makes a few edits to make sure the PMME block has the info it needs to render in the block editor. Prior to this PR, the `wcBlocksCheckoutData` JS variable was added to the page **only when** `WC()->cart` existed - it does not exist in the block editor. This PR makes sure that this variable is always added and uses `WC()->cart` and `WC()->customer` for the values when available, but falls back to defaults when it doesn't.

This PR also conditionally checks for the existence of `window.wcBlocksCheckoutData?.storeCountry` when setting the `currentCountry`. It also only displays the PMME when the amount is greater than `0` and when we have a `currentCountry` set.

image

#### Testing instructions

* Go the edit screen for the block checkout page in wp-admin
* Observe the block error in the Payment block.
* Apply this PR.
* The Payment block should render correctly.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8771` or branch name `fix-stripe-pmme-editor-error` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix-stripe-pmme-editor-error) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 698fb2be28d950e0824f9eca98687b904b7b317a
- Build time: 2024-05-08 13:01:49 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/checkout/blocks/payment-method-label.js [hunk 1]
```diff
@@ -18,23 +18,29 @@ export default ( {
...
+			cartData.totals.currency_minor_unit
+		),
+		10
 	);
 
 	// Customer's country or base country of the store.
 	const currentCountry =
 		cartData.billingAddress.country ||
-		window.wcBlocksCheckoutData.storeCountry;
+		window?.wcBlocksCheckoutData?.storeCountry ||
```
Comment (frosso) [inline client/checkout/blocks/payment-method-label.js]:
```suggestion
		window.wcBlocksCheckoutData?.storeCountry ||
```

Could there be an instance where `window` is not defined?

File: includes/class-wc-payments-blocks-payment-method.php [hunk 2]
```diff
@@ -74,20 +74,18 @@ public function get_payment_method_script_handles() {
...
+			'var wcBlocksCheckoutData = ' . wp_json_encode(
+				[
+					'amount'         => WC()->cart ? WC()->cart->get_total( '' ) : 0,
+					'currency'       => get_woocommerce_currency(),
+					'storeCountry'   => WC()->countries->get_base_country(),
+					'billingCountry' => WC()->customer ? WC()->customer->get_billing_country() : 'US',
+				]
+			) . ';',
+			'before'
+		);
```
Comment (frosso) [inline includes/class-wc-payments-blocks-payment-method.php:88]: It looks like the `wcBlocksCheckoutData` is now added to any editor page that contains a block. But I noticed that on `develop` we're also including some scripts where we technically shouldn't 😬 

It looks like you added some checks on `WC()->cart` and `WC()->customer`, which were my main concern 👍

**Review by frosso — APPROVED** [review]
Just a minor suggestion on the `window?.` conditional - looks good!

File: includes/class-wc-payments-blocks-payment-method.php (hunk 2 above)
Comment (brettshumaker) [reply includes/class-wc-payments-blocks-payment-method.php:88]: Yeah, it's something to look at for sure.

**Pull request merged**